tests/integration/v3_grpc_test.go

```diff
@@ -1952,3 +1952,39 @@ func waitForRestart(t *testing.T, kvc pb.KVClient) {
 		t.Fatalf("timed out waiting for restart: %v", err)
 	}
 }
+
+func TestV3ReadonlyTxnCancelledContext(t *testing.T) {
+	integration.BeforeTest(t)
+	clus := integration.NewCluster(t, &integration.ClusterConfig{
+		Size: 1,
+		// Context should be cancelled in the check that happens inside rangeKeys, but not before
+		GrpcInterceptors: []grpc.UnaryServerInterceptor{injectContextForTxn(newDoneForCallerContext("rangeKeys"))},
+	})
+	defer clus.Terminate(t)
+
+	kvc := integration.ToGRPC(clus.RandClient()).KV
+	// need to insert some data, otherwise rangeKeys will terminate early before checking context
+	pr := &pb.PutRequest{Key: []byte("abc"), Value: []byte("def")}
+	_, err := kvc.Put(context.TODO(), pr)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	txnget := &pb.RequestOp{Request: &pb.RequestOp_RequestRange{RequestRange: &pb.RangeRequest{Key: []byte("abc")}}}
+	txn := &pb.TxnRequest{Success: []*pb.RequestOp{txnget}}
+	_, err = kvc.Txn(context.TODO(), txn)
```

Comment:
I am a little concerned about the readability of the test case. Can you try the following proposal?
1. pass a context with a deadline, i.e. 5s;
2. register an unary interceptor, and just sleep a couple of seconds (5+1=6s) in the interceptor;
3. The `kvc.Txn` should return an error with something like context timeout

Reply:
I am also thinking probably we can just add a unit test to accurately test the different behaviors:
1. For readonly transaction, it should NOT panic due to context timeout;
2. For non-readonly trancation, it should panic if the context timeout

Reply:
Deadline approach doesn't work, because it's early terminated [here](https://github.com/etcd-io/etcd/blob/main/server/etcdserver/v3_server.go#L170) (tried it out).

Unit tests are fine with me. I've mentioned this [earlier](https://github.com/etcd-io/etcd/pull/14149#discussion_r920533239). 
> Another approach I see is to add test only on txn.Txn level and not replicate panic as integration test. Maybe not ideal, but integrations tests don't have to cover every failure scenario.

Will work on PR to convert to unit tests and remove this test.